### DISCUSSION TITLE
Cloud provisioner

### DISCUSSION THREAD

**Discussion by smalchow** [smallstep/certificates#668] [Q&A] [answered]
Dears,

we are thinking about using smallstep, however, we have a question about the cloud provisioners.
To work, it looks as if the /provisioner endpoint needs to be open, otherwise the step client can not authenticate with its IID.
First question: is that correct or can we do it differently ?

If the endpoint is open, this means, that anyone can retrieve a list of provisioners without any authentication, which would also mean a list of cloud provider account numbers, right ? 

This would pose a security problem

Cheers
Soeren

**Comment by maraino** [comment]
Hi @smalchow, this is right, the provisioners list is public, and it can provide some kind of risk. In most of cases, you will see only public keys/certs or encrypted private keys (JWK). On OIDC provisioners you will have access to the client_secret, but with a properly configured IdP, this shouldn't be a security risk as you will need to log in first.

There're multiple workarounds. The most obvious case is when the CA is in a private network, and `/provisioners` is only available inside the network, with optional firewall rules in place. VPN is also an obvious one. 

But another solution that we've proposed, and have `step`/`step-ca` supports it the use of TLS-over-mTLS tunnels using tools like [Ghostunnel](https://github.com/ghostunnel/ghostunnel) or even some configurations of Nginx or other proxies. We generally use these tunnels to talk between RAs and CAs, when the CA is in the cloud, and the RA is in an internal network.

We don't provide documentation for this setup, but the PR that added support for this explains how to set up this using Ghostunnel and Nginx https://github.com/smallstep/certificates/pull/531

The implementation is intended to work between RAs, and CAs, the `step` client has limited support for it, but with some gotchas, it can be used.

A similar workaround can be used with a proxy that requires mTLS authentication, but in this case we're not doing TLS-over-mTLS, the major problem here is the bootstrap of the first certificate, if you don't have access to the `/provisioners` endpoint. But to be short, an identity.json in `$(step path)/config/identity.json` like this one will use mTLS to talk to the CA in most cases:
```json
{
   "type": "mTLS",
   "crt": "/home/mariano/.step/identity/identity.crt",
   "key": "/home/mariano/.step/identity/identity_key"
}
```

But it's a challenge to bootstrap this identity and keep it alive. 

We have other ideas in mind to make this easy, but right now I don't have a roadmap that I can give you.

**Comment by mikemaxey** [comment]
@smalchow Maxey, VP of Product here. We discussed this as a team today. We have a few ways to solve this depending on your end goal and use case. Can we speak live so we can make a better recommendation? Ping me at maxey at smallstep.com.

**Comment by joaopedrolourencoaffonso** [comment]
# So Smallstep shouldn't be public?

Correct if I am wrong, but putting Smallstep in a public enviroment, with a public IP, is therefore non-viable, as will allow anyone who knows how to use the APIs to sign it's own certificate. Correct?

**Reply by maraino** [reply]
You can use step-ca in a public IP, but you need to be careful and add some limitations, for example, if using an OIDC provisioner you should limit the domain, using the `domains` property. For GCP you can use `serviceAccounts` and `projectIDs`. For AWS `accounts` and for Azure `tenantId` and `resourceGroups`.

But there is public information that would be available, like the provisioner configurations or your roots certificates. They are required for bootstrapping.

You can also do TLS-over-mTLS to avoid that, but that's a more complex configuration, see https://github.com/smallstep/certificates/pull/531

**Reply by joaopedrolourencoaffonso** [reply]
My only provisioner is myself throught an API that I made for authentication with Telegram.

Before trying #531 , I wanted to know. The local provisioner private key is encrypted on ca.json, in other words, it's as safe as the password for it, which, obviously, can be broken with brute-force. 

I was thinking about creating a script to edit the "ca.json" so to substitute the password of the encrypted key (every one hour, for example), this wouldn't affect the working of the certificates nor step-ca.  My questions are:

1) Would this work?
2) The provisioner private key stored on the ca.json is stored unecrypted somewhere?
3) In case 2 is negative, can I use ```step crypto``` or other tool to decrypt the provisioner key? If not, what is the algorithm being used? So I can search for it.

Another solution that I thought of is:
```
step ca token internal.smallstep.com --key token.key
```
I could use a private key the only the registration.py has and configure a corresponding public key for the CA. The only thing is that I still can't find on the documentation *how* I can do this. If you could point me I would be very gratefull.

If none these options work, then I think that I will try the #531 and if _that_ fails, so I will just put  my CA back on the private network and use registration.py as a proxy for it.

**Reply by joaopedrolourencoaffonso** [reply]
> But there is public information that would be available, like the provisioner configurations or your roots certificates. They are required for bootstrapping

In the case of my system, all applications would came with these pre-installed, so it would be a little redundant, but I got the point and will see how to explore it.

**Reply by joaopedrolourencoaffonso** [reply]
I was digging the step ca provisioner and found [this](https://smallstep.com/docs/step-cli/reference/ca/provisioner/add) page with this example of using a Identity Provisioner based on X5C certificates. I maybe wrong but I think that this is related to the: ```step ca token internal.smallstep.com --key token.key``` command above correct?

If this is the case, could I simply add my a provisioner authority with the address of the registration.py API, and the make my API generate sign it's own tokens with it's own private key?

I maybe wrong, but in this case, the only public thing would be the root certificate of correspondent to ```registration.py``` private key correct?

**Reply by maraino** [reply]
> Before trying #531 , I wanted to know. The local provisioner private key is encrypted on ca.json, in other words, it's as safe as the password for it, which, obviously, can be broken with brute-force.

This is generally the case, but the encrypted key is not required, you can remove it from the ca.json. And as you mention you can use `step ca token --key token.key internal.smallstep.com`.

> I was thinking about creating a script to edit the "ca.json" so to substitute the password of the encrypted key (every one hour, for example), this wouldn't affect the working of the certificates nor step-ca. My questions are:
> Would this work?

Bruteforce on a good password is not viable.

> The provisioner private key stored on the ca.json is stored unecrypted somewhere?

No 

> In case 2 is negative, can I use step crypto or other tool to decrypt the provisioner key? If not, what is the algorithm being used? So I can search for it.

You can use `step` for it:
```sh
echo "myencryptedkey" | step crypto jwe decrypt
```

> I was digging the step ca provisioner and found this page with this example of using a Identity Provisioner based on X5C certificates. I maybe wrong but I think that this is related to the: step ca token internal.smallstep.com --key token.key command above correct?

With X5C is very similar to `--key` but in this case is with `--x5c-cert` and `--x5c-key`:

```sh
step ca token --x5c-cert x5c.crt --x5c-key x5c.key internal.smallstep.com
```

> If this is the case, could I simply add my a provisioner authority with the address of the registration.py API, and the make my API generate sign it's own tokens with it's own private key?

You can do that with X5C or JWK too (removing the key from the ca.json)

> I maybe wrong, but in this case, the only public thing would be the root certificate of correspondent to registration.py private key correct?

Yes.

**Reply by joaopedrolourencoaffonso** [reply]
1 - Thank you a lot! This was superhelpful!

2 -
>This is generally the case, but the encrypted key is not required, you can remove it from the ca.json. And as you mention you can use step ca token --key token.key internal.smallstep.com.

Ok, so I can simply cut the encrypted key from ca.json, put it on a token.key file and use it as normal, without risk to my CA, correct?

**Comment by joaopedrolourencoaffonso** [comment]
# How to use X5C identity provider?

So, I am having a problem. For solving the security problem of letting the step-ca on a public IP, I am trying to configure my own X5C identity provisioner, so I used this command to generate a root certificate:
```bash
step certificate create registration registration_crt.pem registration_key.pem --profile root-ca --kty RSA --size 4096
```
Then, I added the new identity provisioner with this command:
```bash
step ca provisioner add registration --type X5C --x5c-root registration_crt.pem
```
But when I tried to get a signed token from step, I got this error:
```bash
joao@joao-VirtualBox:~/Desktop/smallstep_provisioner$ step ca token joao_teste --x5c-cert registration_crt.pem --x5c-key registration_key.pem --password-file pass.txt 

✔ Provisioner: registration (X5C) 

error validating x5c certificate chain and key for use in x5c header: ValidateX5C: certificate/private-key pair used to sign token is not approved for digital signature
```
Then I tried to use [this](https://smallstep.com/docs/step-ca/provisioners#x5c-x509-certificate) example:
```
joao@joao-VirtualBox:~/Desktop/smallstep_provisioner$ step ca provisioner add x5c-smallstep --type x5c --x5c-root $(step path)/certs/root_ca.crt

joao@joao-VirtualBox:~/Desktop/smallstep_provisioner$ step ca token teste --provisioner x5c-smallstep --provisioner-password-file pass.txt --x5c-cert root_ca.crt --x5c-key root_ca_key --password-file pass.txt

✔ Provisioner: x5c-smallstep (X5C)

error validating x5c certificate chain and key for use in x5c header: ValidateX5C: certificate/private-key pair used to sign token is not approved for digital signature
``` 
What does this error mean? What would be an "approved" certificate/key pair? How can I make one? I have search the discussions and the issues and couldn't find nothing beyond an gitter page that is  no longer on the air.

Any help is welcome!

**Reply by maraino** [reply]
You need to use a leaf cert signed by the root that you created. An easy way to try it is by configuring an X5C provisioner with the CA root. Then you can use any other provisioner to grant a certificate that you can use on that X5C provisioner.

With a separate root:
```
## ROOT
$ step certificate create registration registration_crt.pem registration_key.pem --profile root-ca --kty RSA --size 4096

## Optionally an intermediate
$ step certificate create intermediate intermediate_crt.pem intermediate_key.pem --profile intermediate-ca --kty RSA --size 4096 --ca registration_crt.pem --ca-key registration_key.pem

## Leaf (do not use --bundle if you don't have an intermediate)
$ step certificate create intermediate leaf_crt.pem leaf_key.pem --kty RSA --size 4096 --ca intermediate_crt.pem --ca-key intermediate_key.pem --bundle
```

**Reply by joaopedrolourencoaffonso** [reply]
Ohhh! I got it, I need to provide an bundle to create the leaf certificate, and **then**, use: ```step ca provisioners add```. I am going to experiment rigth now!

Thanks for the speed!

**Comment by joaopedrolourencoaffonso** [comment, marked as answer]
Hi! I just wanted to inform that I tried both ways and both worked! Later, I will make a post on the [directory](https://github.com/joaopedrolourencoaffonso/python_smallstep/) explaining everything in details.

Thanks a lot @maraino, you're a life saver!
